### PR TITLE
Add Negative Precision Support for JS

### PR DESCRIPTION
**Pull request by adi90x**
Hello @kroitor , 
Quick PR  to implement Negative precision for JS Base.
It already exist for php & Python thanks to @frosty00 
It is link to PR #3741  & issue #3847 
What do you think ?

### CODE REVIEW

**Comment by kroitor** [comment]
Hi, @adi90x ! First of, thx for the contribution!

> What do you think ?

I think we will need to uncomment the following lines and add some more extensive tests for negative cases here: 

https://github.com/ccxt/ccxt/blob/master/js/test/base/functions/test.number.js#L84

If you uncomment that and run that test against the edited code it should show up.

Unfortunately, the code you've submitted doesn't work on my side:

```
> log.green ((Math.pow(10,-2)*ccxt.decimalToPrecision (123.456/Math.pow(10,-2),ccxt.ROUND,0)).toString())
123.46000000000001 (repl:1)
'123.46000000000001'
```

**Comment by kroitor** [comment]
Sorry, there's actually two such groups of tests in that file:

- https://github.com/ccxt/ccxt/blob/master/js/test/base/functions/test.number.js#L32
- https://github.com/ccxt/ccxt/blob/master/js/test/base/functions/test.number.js#L84

@adi90x can you make it pass the tests?

**Comment by adi90x** [comment]
I guess it failed on TRUNCATED test, I am looking into it !

**Comment by adi90x** [comment]
It should pass the test now , btw one test was not correct so I update it

**Comment by frosty00** [comment]
@adi90x thanks for trying to solve this by no means easy task.

Can you copy and paste the follow tests into testnumber.js and verify they pass?

```
assert (decimalToPrecision ('5', ROUND, -1, DECIMAL_PLACES) === '10');
assert (decimalToPrecision ('4.999', ROUND, -1, DECIMAL_PLACES) === '0');
assert (decimalToPrecision ('0.0431531423', ROUND, -1, DECIMAL_PLACES) === '0');
assert (decimalToPrecision ('-69.3', ROUND, -1, DECIMAL_PLACES) === '-70');
assert (decimalToPrecision ('5001', ROUND, -4, DECIMAL_PLACES) === '10000');
assert (decimalToPrecision ('4999.999', ROUND, -4, DECIMAL_PLACES) === '0');
assert (decimalToPrecision ('69.3', TRUNCATE, -2, DECIMAL_PLACES) === '100');
assert (decimalToPrecision ('-69.3', TRUNCATE, -2, DECIMAL_PLACES) === '-100');
assert (decimalToPrecision ('69.3', TRUNCATE, -1, SIGNIFICANT_DIGITS) === '60');
assert (decimalToPrecision ('-69.3', TRUNCATE, -1, SIGNIFICANT_DIGITS) === '-60');
assert (decimalToPrecision ('69.3', TRUNCATE, -2, SIGNIFICANT_DIGITS) === '100');
```

**Comment by adi90x** [comment]
Hello,

Don't really agree on those , their are OK if we use ROUND ,  but if we TRUNCATE 69.3 at -2 precision it should be 0 and not 100 , no ?

> ```
> assert (decimalToPrecision ('69.3', TRUNCATE, -2, DECIMAL_PLACES) === '100');
> assert (decimalToPrecision ('-69.3', TRUNCATE, -2, DECIMAL_PLACES) === '-100');
> assert (decimalToPrecision ('69.3', TRUNCATE, -2, SIGNIFICANT_DIGITS) === '100');
> ```

**Comment by frosty00** [comment]
No because 69.3 is closer to 100 than 0

I think the behaviour we are trying to get is simillar to python's `round`:

```
round(69, -2)
Out[30]: 100
```

**Comment by adi90x** [comment]
@frosty00  make sense for rounding , but in case of truncate if shouldn't , otherwise there is no difference between truncate and rounding...

**Comment by frosty00** [comment]
Can you give me an example?

**Comment by adi90x** [comment]
Let's say 99 truncate to one negative significant number is 90 even if it is closer to 100, and rounded it is 100.
In the same way , 9.99 truncate to one "positive" number is 9.9 (even if it is closer to 10 ) and rounded it is at 10 .
No ?

**Comment by adi90x** [comment]
Not sure it has a lot a meaning to truncate, that is why I have try to implement it as the opposite of the truncate with positive number ( ie : keeping only digits over a certains level )
Last commit , should be fine with the same implementation for js,php and python.

**Comment by frosty00** [comment]
Can you provide examples of what it means to truncate a number with negative precision in the same format I provided in the example above.

**Comment by frosty00** [comment]
@adi90x now I have to change my implementations in python and php 0:

**Comment by adi90x** [comment]
> @adi90x now I have to change my implementations in python and php 0:

I have update python & php , just struggling a bit with linting I think

**Comment by frosty00** [comment]
Actually on second thought I was right all along, that's just the same as positive truncation, surely the solution to this is actually really simple.

We can just return `decimalToPrecision (n, TRUNCATE, |prec|, counting_mode, padding_mode)`

where |prec| is the absolute value of the precision.

**Comment by adi90x** [comment]
Not really , 
As trunc(123.456 , -2 )  should give 100 and `decimalToPrecision (n, TRUNCATE, |prec|, counting_mode, padding_mode)` would give 123.45.

**Comment by frosty00** [comment]
Can you provide examples in the same format as I did [here](https://github.com/ccxt/ccxt/pull/4225#issuecomment-441453089)

**Comment by adi90x** [comment]
> Can you provide examples in the same format as I did [here](https://github.com/ccxt/ccxt/pull/4225#issuecomment-441453089)

Would see it like that : 
ROUND
55.3 -1 -> 60
55.3 -2 -> 100
55.3 -3 -> 0

TRUNCATE
55.3 -1 -> 50
55.3 -2 -> 0
55.3 -3 -> 0

This is to have : 
99.99 -1 -> 90
99.99 -2 -> 0
like when we truncate with positive number  : 99.99 1 -> 99.9  <<==>>  99.99 -1 -> 90

**Comment by frosty00** [comment]
Ok I understand. 

Can you make a new branch with only the javascript commits.

```
git add 
git commit
git checkout origin/master 
git checkout -b javascriptdectoprec
git checkout implementNegPrecisionJS js/base/functions/number.js js/test/base/functions/test.number.js
```

and push that branch instead because I'm worried you will add bugs to the py/php files, and it makes it easier to merge later

**Comment by frosty00** [comment]
also the solution in python to what you described is:

```
int(str(n)[:prec] + '0' *  abs(prec))
```

**Comment by adi90x** [comment]
> also the solution in python to what you described is:
> 
> ```
> int(str(n)[:prec] + '0' *  abs(prec))
> ```

Thanks a lot for the solution ! don't like python at all  :p 
Will do a jsonly branch in a second !

**Comment by adi90x** [comment]
Push another PR #4226 with only JS if you want to aggregate everything ?

**Comment by frosty00** [comment]
@adi90x patience, this probably won't get merged today. Let me implement python and php first...

**Comment by adi90x** [comment]
@frosty00  tell me what you think it should be fine now,yo I have usu don't have anything to do :) I have use the same approach on all 3 language ( recalling decimal_to_precision ) , but you can reimplement something quicker if you prefer ?

**Comment by kroitor** [comment]
@adi90x can you plz exclude the following files from this PR (restore from master):

     
- build/ccxt.browser.js
- doc/README.rst
- doc/exchanges-by-country.rst
- doc/exchanges.rst
- doc/manual.rst
- python/README.rst

? They are generated automatically, so we don't normally add them... Not a very big deal, it would just make browsing the actual edits easier.

**Comment by adi90x** [comment]
@kroitor files removed ! 
I definitly need to find a way to be sure to avoid push does next time.
On this PR , two possibilities : here I have rewrote php & python to match what I did on JS.
On the other PR, it was only the javascript part if you prefer to keep @frosty00 approch on the other language.

**Comment by frosty00** [comment]
Nice, your recursive approach is really clever and cleaner than mine. I think we could merge this branch instead and I won't need to edit the py/php files. 

Thanks again @adi90x

File: python/ccxt/base/decimal_to_precision.py [hunk 1]
```diff
@@ -41,12 +42,22 @@ def decimal_to_precision(n, rounding_mode=ROUND, precision=None, counting_mode=D
...
 
     def power_of_10(x):
         return decimal.Decimal('10') ** (-x)
 
+    if precision < 0:
+        to_nearest = power_of_10(precision)
+        if rounding_mode == ROUND:
+            return str(to_nearest * decimal.Decimal(decimal_to_precision(dec / to_nearest, rounding_mode, 0, DECIMAL_PLACES, padding_mode)))
+        elif rounding_mode == TRUNCATE:
+            return str(decimal_to_precision(dec - dec % to_nearest, rounding_mode, 0, DECIMAL_PLACES, padding_mode))
```
Comment (frosty00) [inline python/ccxt/base/decimal_to_precision.py]: you can remove this str() call since decimal_to_precision already returns a string

**Comment by frosty00** [comment]
Also if you run `npm run build` you can run the tests locally

File: python/ccxt/base/decimal_to_precision.py (hunk 1 above)
Comment (adi90x) [reply python/ccxt/base/decimal_to_precision.py]: Removed !

**Comment by adi90x** [comment]
@kroitor @frosty00 completly fuck up everything on this PR moving clean code in to PR  #4226  and closing this one.

**Comment by kroitor** [comment]
Merged in #4226! Thank you for the effort, @adi90x ! We really appreciate it!
